### PR TITLE
implicit parameter casting in function calls 

### PR DESCRIPTION
**Pull request by lankavitharana**

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 1]
```diff
@@ -1951,6 +1959,86 @@ private void linkFunction(FunctionInvocationExpr funcIExpr) {
...
+                if (newExpr != null) {
+                    exprs[i] = newExpr;
+                } else {
+                    implicitCastPossible = false;
+                    break;
+                }
+            }
+            if (implicitCastPossible && functionSymbol == null) {
+                functionSymbol = (BLangSymbol) entry.getValue();
+            } else {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: There is a bug in this code.  Assume I want to invoke the function "print(float)", but only the following functions are defined in the symbol map.

print(string)
print(int)
print(Person)
print(Employee)

This code will try to throw an "ambiguous function error ".

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/FunctionSymbolName.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+*  software distributed under the License is distributed on an
+*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+*  KIND, either express or implied.  See the License for the
+*  specific language governing permissions and limitations
+*  under the License.
+*/
+package org.ballerinalang.model;
+
+/**
+ * {@code FunctionSymbolName} represents a package qualified name of a {@link Symbol} in Ballerina.
```
Comment (lankavitharana) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/FunctionSymbolName.java]: done

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/FunctionSymbolName.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    public int getNoOfParameters() {
+        return noOfParameters;
+    }
+
+    public String getFuncName() {
+        return funcName;
+    }
+
+    public boolean isFunctionsEqual(FunctionSymbolName funcSymName) {
```
Comment (lankavitharana) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/FunctionSymbolName.java]: done

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java (hunk 1 above)
Comment (lankavitharana) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: fixed this and added a test case as well

**Pull request merged**